Split hi res geocoding

rafatower opened the pull request:
Just to get an idea of how many tests I broke

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4694/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4695/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4733/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4737/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4746/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4760/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4761/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4767/))

Cartofante commented:
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4777/))

rafatower commented:
What's here:
- Split the former `Geocoder` (here) into `HiresGeocoder` and `HiresBatchGeocoder`: now they know about the service and API they work with but not much about db's, tables or the rest.
- A new class `HiresGeocoderFactory` to decide whether to instantiate one or another.
- Many changes to the `TableGeocoder` to accommodate the above. E.g: moved the state polling loop on the here service to the `HiresBatchGeocoder`, so from the client classes that's kind of transparent.
- Stuff from public to protected/private
- Fixes for legacy tests, now turned into kind of white box (accessing their privates :stuck_out_tongue_winking_eye: )
- Many small details, bugfixes and some TODO's.

Some comments about this:
- I'd like to add more tests, they're easier now but this task will have to be "time boxed" so I'll do my best.
- I have not thrown away the existing tests yet because they proved to be useful to check against expected behavior, though ideally they're to be deleted.
- Please don't be too hard on it, I don't like remaining TODO's there but again I tried to carry on by priority and the TODO's let me capture things in context that otherwise are lost in wiki pages/tickets...

@Kartones and / or @juanignaciosl please review

cc/ @ethervoid so that you can get familiar with the code. Don't get too scared :smile:

juanignaciosl commented:
LGTM :+1: :clap:

Kartones commented on services/geocoder/lib/hires_batch_geocoder.rb:
```diff
@@ -0,0 +1,167 @@
+    end
+
+    def upload
+      assert_batch_api_enabled
```
As we discussed in person and you said would be better at the factory, but at least a good starting point!

Kartones commented:
:+1:
